From 3c8e4852477d5b1e2da877808c998dc57db9460f Mon Sep 17 00:00:00 2001
From: Andrey Kosyakov <caseq@chromium.org>
Date: Tue, 23 Jan 2018 05:20:27 +0000
Subject: [PATCH] DevTools: speculative fix for crash in
 NetworkHandler::Disable

This keeps BrowserContext* and StoragePartition* instead of
RenderProcessHost* in an attemp to resolve UAF of RenderProcessHost
upon closure of DevTools front-end.

Bug: 801117, 783067, 780694
Change-Id: I6c2cca60cc0c29f0949d189cf918769059f80c1b
Reviewed-on: https://chromium-review.googlesource.com/876657
Commit-Queue: Andrey Kosyakov <caseq@chromium.org>
Reviewed-by: Dmitry Gozman <dgozman@chromium.org>
Cr-Commit-Position: refs/heads/master@{#531157}
---
 content/browser/devtools/devtools_session.cc  | 16 +++--
 content/browser/devtools/devtools_session.h   |  6 +-
 .../protocol/devtools_domain_handler.cc       |  2 +-
 .../protocol/devtools_domain_handler.h        |  3 +-
 .../browser/devtools/protocol/dom_handler.cc  |  2 +-
 .../browser/devtools/protocol/dom_handler.h   |  2 +-
 .../devtools/protocol/emulation_handler.cc    |  2 +-
 .../devtools/protocol/emulation_handler.h     |  2 +-
 .../devtools/protocol/input_handler.cc        |  2 +-
 .../browser/devtools/protocol/input_handler.h |  2 +-
 .../devtools/protocol/inspector_handler.cc    |  2 +-
 .../devtools/protocol/inspector_handler.h     |  2 +-
 .../browser/devtools/protocol/io_handler.cc   | 22 +++---
 .../browser/devtools/protocol/io_handler.h    |  7 +-
 .../devtools/protocol/network_handler.cc      | 71 +++++++++----------
 .../devtools/protocol/network_handler.h       |  8 ++-
 .../browser/devtools/protocol/page_handler.cc |  2 +-
 .../browser/devtools/protocol/page_handler.h  |  2 +-
 .../devtools/protocol/security_handler.cc     |  2 +-
 .../devtools/protocol/security_handler.h      |  2 +-
 .../protocol/service_worker_handler.cc        | 33 ++++-----
 .../protocol/service_worker_handler.h         |  7 +-
 .../devtools/protocol/storage_handler.cc      | 39 +++++-----
 .../devtools/protocol/storage_handler.h       |  6 +-
 .../devtools/protocol/target_handler.cc       |  2 +-
 .../devtools/protocol/target_handler.h        |  2 +-
 .../render_frame_devtools_agent_host.cc       |  5 +-
 .../service_worker_devtools_agent_host.cc     | 15 ++--
 .../shared_worker_devtools_agent_host.cc      | 16 ++---
 .../shared_worker_devtools_agent_host.h       |  2 -
 30 files changed, 144 insertions(+), 142 deletions(-)

diff --git a/content/browser/devtools/devtools_session.cc b/content/browser/devtools/devtools_session.cc
index d5f9bc4bc6551d..cf6e07122f5c16 100644
--- a/content/browser/devtools/devtools_session.cc
+++ b/content/browser/devtools/devtools_session.cc
@@ -11,6 +11,7 @@
 #include "content/browser/devtools/render_frame_devtools_agent_host.h"
 #include "content/browser/frame_host/render_frame_host_impl.h"
 #include "content/public/browser/devtools_manager_delegate.h"
+#include "content/public/common/child_process_host.h"
 
 namespace content {
 
@@ -33,7 +34,7 @@ DevToolsSession::DevToolsSession(DevToolsAgentHostImpl* agent_host,
     : binding_(this),
       agent_host_(agent_host),
       client_(client),
-      process_(nullptr),
+      process_host_id_(ChildProcessHost::kInvalidUniqueID),
       host_(nullptr),
       dispatcher_(new protocol::UberDispatcher(this)),
       weak_factory_(this) {
@@ -50,16 +51,16 @@ DevToolsSession::~DevToolsSession() {
 void DevToolsSession::AddHandler(
     std::unique_ptr<protocol::DevToolsDomainHandler> handler) {
   handler->Wire(dispatcher_.get());
-  handler->SetRenderer(process_, host_);
+  handler->SetRenderer(process_host_id_, host_);
   handlers_[handler->name()] = std::move(handler);
 }
 
-void DevToolsSession::SetRenderer(RenderProcessHost* process_host,
+void DevToolsSession::SetRenderer(int process_host_id,
                                   RenderFrameHostImpl* frame_host) {
-  process_ = process_host;
+  process_host_id_ = process_host_id;
   host_ = frame_host;
   for (auto& pair : handlers_)
-    pair.second->SetRenderer(process_, host_);
+    pair.second->SetRenderer(process_host_id_, host_);
 }
 
 void DevToolsSession::SetBrowserOnly(bool browser_only) {
@@ -215,9 +216,10 @@ void DevToolsSession::DispatchProtocolMessage(
 
 void DevToolsSession::ReceivedBadMessage() {
   MojoConnectionDestroyed();
-  if (process_) {
+  RenderProcessHost* process = RenderProcessHost::FromID(process_host_id_);
+  if (process) {
     bad_message::ReceivedBadMessage(
-        process_, bad_message::RFH_INCONSISTENT_DEVTOOLS_MESSAGE);
+        process, bad_message::RFH_INCONSISTENT_DEVTOOLS_MESSAGE);
   }
 }
 
diff --git a/content/browser/devtools/devtools_session.h b/content/browser/devtools/devtools_session.h
index 0916da16c83f04..72d19e444b564e 100644
--- a/content/browser/devtools/devtools_session.h
+++ b/content/browser/devtools/devtools_session.h
@@ -35,10 +35,8 @@ class DevToolsSession : public protocol::FrontendChannel,
   void SetBrowserOnly(bool browser_only);
 
   void AddHandler(std::unique_ptr<protocol::DevToolsDomainHandler> handler);
-
   // TODO(dgozman): maybe combine this with AttachToAgent?
-  void SetRenderer(RenderProcessHost* process_host,
-                   RenderFrameHostImpl* frame_host);
+  void SetRenderer(int process_host_id, RenderFrameHostImpl* frame_host);
 
   void AttachToAgent(const blink::mojom::DevToolsAgentAssociatedPtr& agent);
   void DispatchProtocolMessage(const std::string& message);
@@ -89,7 +87,7 @@ class DevToolsSession : public protocol::FrontendChannel,
   bool browser_only_ = false;
   base::flat_map<std::string, std::unique_ptr<protocol::DevToolsDomainHandler>>
       handlers_;
-  RenderProcessHost* process_;
+  int process_host_id_;
   RenderFrameHostImpl* host_;
   std::unique_ptr<protocol::UberDispatcher> dispatcher_;
 
diff --git a/content/browser/devtools/protocol/devtools_domain_handler.cc b/content/browser/devtools/protocol/devtools_domain_handler.cc
index db31f7d82dfdeb..5290245b0f7d9c 100644
--- a/content/browser/devtools/protocol/devtools_domain_handler.cc
+++ b/content/browser/devtools/protocol/devtools_domain_handler.cc
@@ -17,7 +17,7 @@ DevToolsDomainHandler::DevToolsDomainHandler(const std::string& name)
 DevToolsDomainHandler::~DevToolsDomainHandler() {
 }
 
-void DevToolsDomainHandler::SetRenderer(RenderProcessHost* process_host,
+void DevToolsDomainHandler::SetRenderer(int process_host_id,
                                         RenderFrameHostImpl* frame_host) {}
 
 void DevToolsDomainHandler::Wire(UberDispatcher* dispatcher) {
diff --git a/content/browser/devtools/protocol/devtools_domain_handler.h b/content/browser/devtools/protocol/devtools_domain_handler.h
index 8d8b4f1271f024..ea48934a0f1a86 100644
--- a/content/browser/devtools/protocol/devtools_domain_handler.h
+++ b/content/browser/devtools/protocol/devtools_domain_handler.h
@@ -10,7 +10,6 @@
 namespace content {
 
 class RenderFrameHostImpl;
-class RenderProcessHost;
 
 namespace protocol {
 
@@ -19,7 +18,7 @@ class DevToolsDomainHandler {
   explicit DevToolsDomainHandler(const std::string& name);
   virtual ~DevToolsDomainHandler();
 
-  virtual void SetRenderer(RenderProcessHost* process_host,
+  virtual void SetRenderer(int process_host_id,
                            RenderFrameHostImpl* frame_host);
   virtual void Wire(UberDispatcher* dispatcher);
   virtual Response Disable();
diff --git a/content/browser/devtools/protocol/dom_handler.cc b/content/browser/devtools/protocol/dom_handler.cc
index 9bf4e6eead8a83..2fc40a76ff08bf 100644
--- a/content/browser/devtools/protocol/dom_handler.cc
+++ b/content/browser/devtools/protocol/dom_handler.cc
@@ -24,7 +24,7 @@ void DOMHandler::Wire(UberDispatcher* dispatcher) {
   DOM::Dispatcher::wire(dispatcher, this);
 }
 
-void DOMHandler::SetRenderer(RenderProcessHost* process_host,
+void DOMHandler::SetRenderer(int process_host_id,
                              RenderFrameHostImpl* frame_host) {
   host_ = frame_host;
 }
diff --git a/content/browser/devtools/protocol/dom_handler.h b/content/browser/devtools/protocol/dom_handler.h
index 0dfcd947b1130b..7c6bbc6deff61f 100644
--- a/content/browser/devtools/protocol/dom_handler.h
+++ b/content/browser/devtools/protocol/dom_handler.h
@@ -22,7 +22,7 @@ class DOMHandler : public DevToolsDomainHandler,
   ~DOMHandler() override;
 
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
   Response Disable() override;
 
diff --git a/content/browser/devtools/protocol/emulation_handler.cc b/content/browser/devtools/protocol/emulation_handler.cc
index b9b4992812411e..3aa6d5141cacde 100644
--- a/content/browser/devtools/protocol/emulation_handler.cc
+++ b/content/browser/devtools/protocol/emulation_handler.cc
@@ -64,7 +64,7 @@ EmulationHandler::EmulationHandler()
 EmulationHandler::~EmulationHandler() {
 }
 
-void EmulationHandler::SetRenderer(RenderProcessHost* process_host,
+void EmulationHandler::SetRenderer(int process_host_id,
                                    RenderFrameHostImpl* frame_host) {
   if (host_ == frame_host)
     return;
diff --git a/content/browser/devtools/protocol/emulation_handler.h b/content/browser/devtools/protocol/emulation_handler.h
index 423d18f07a0c3d..f322802b6481d5 100644
--- a/content/browser/devtools/protocol/emulation_handler.h
+++ b/content/browser/devtools/protocol/emulation_handler.h
@@ -24,7 +24,7 @@ class EmulationHandler : public DevToolsDomainHandler,
   ~EmulationHandler() override;
 
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
 
   Response Disable() override;
diff --git a/content/browser/devtools/protocol/input_handler.cc b/content/browser/devtools/protocol/input_handler.cc
index 97700f659c39b7..92c708965a5d61 100644
--- a/content/browser/devtools/protocol/input_handler.cc
+++ b/content/browser/devtools/protocol/input_handler.cc
@@ -267,7 +267,7 @@ std::vector<InputHandler*> InputHandler::ForAgentHost(
       host, Input::Metainfo::domainName);
 }
 
-void InputHandler::SetRenderer(RenderProcessHost* process_host,
+void InputHandler::SetRenderer(int process_host_id,
                                RenderFrameHostImpl* frame_host) {
   if (frame_host == host_)
     return;
diff --git a/content/browser/devtools/protocol/input_handler.h b/content/browser/devtools/protocol/input_handler.h
index 38638ec0d96fce..6e0d79fc7df9f5 100644
--- a/content/browser/devtools/protocol/input_handler.h
+++ b/content/browser/devtools/protocol/input_handler.h
@@ -37,7 +37,7 @@ class InputHandler : public DevToolsDomainHandler,
   static std::vector<InputHandler*> ForAgentHost(DevToolsAgentHostImpl* host);
 
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
 
   void OnSwapCompositorFrame(
diff --git a/content/browser/devtools/protocol/inspector_handler.cc b/content/browser/devtools/protocol/inspector_handler.cc
index 480277aec34f54..8d92ab68ec6889 100644
--- a/content/browser/devtools/protocol/inspector_handler.cc
+++ b/content/browser/devtools/protocol/inspector_handler.cc
@@ -30,7 +30,7 @@ void InspectorHandler::Wire(UberDispatcher* dispatcher) {
   Inspector::Dispatcher::wire(dispatcher, this);
 }
 
-void InspectorHandler::SetRenderer(RenderProcessHost* process_host,
+void InspectorHandler::SetRenderer(int process_host_id,
                                    RenderFrameHostImpl* frame_host) {
   host_ = frame_host;
 }
diff --git a/content/browser/devtools/protocol/inspector_handler.h b/content/browser/devtools/protocol/inspector_handler.h
index 716364be6c8a8b..d41964199fd34e 100644
--- a/content/browser/devtools/protocol/inspector_handler.h
+++ b/content/browser/devtools/protocol/inspector_handler.h
@@ -26,7 +26,7 @@ class InspectorHandler : public DevToolsDomainHandler,
       DevToolsAgentHostImpl* host);
 
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
 
   void TargetCrashed();
diff --git a/content/browser/devtools/protocol/io_handler.cc b/content/browser/devtools/protocol/io_handler.cc
index 9d3c0de4ae0efc..e01d5f29a13f5f 100644
--- a/content/browser/devtools/protocol/io_handler.cc
+++ b/content/browser/devtools/protocol/io_handler.cc
@@ -25,7 +25,8 @@ namespace protocol {
 IOHandler::IOHandler(DevToolsIOContext* io_context)
     : DevToolsDomainHandler(IO::Metainfo::domainName),
       io_context_(io_context),
-      process_host_(nullptr),
+      browser_context_(nullptr),
+      storage_partition_(nullptr),
       weak_factory_(this) {}
 
 IOHandler::~IOHandler() {}
@@ -35,9 +36,16 @@ void IOHandler::Wire(UberDispatcher* dispatcher) {
   IO::Dispatcher::wire(dispatcher, this);
 }
 
-void IOHandler::SetRenderer(RenderProcessHost* process_host,
+void IOHandler::SetRenderer(int process_host_id,
                             RenderFrameHostImpl* frame_host) {
-  process_host_ = process_host;
+  RenderProcessHost* process_host = RenderProcessHost::FromID(process_host_id);
+  if (process_host) {
+    browser_context_ = process_host->GetBrowserContext();
+    storage_partition_ = process_host->GetStoragePartition();
+  } else {
+    browser_context_ = nullptr;
+    storage_partition_ = nullptr;
+  }
 }
 
 void IOHandler::Read(
@@ -50,15 +58,13 @@ void IOHandler::Read(
 
   scoped_refptr<DevToolsIOContext::ROStream> stream =
       io_context_->GetByHandle(handle);
-  if (!stream && process_host_ &&
+  if (!stream && browser_context_ &&
       StartsWith(handle, kBlobPrefix, base::CompareCase::SENSITIVE)) {
-    BrowserContext* browser_context = process_host_->GetBrowserContext();
     ChromeBlobStorageContext* blob_context =
-        ChromeBlobStorageContext::GetFor(browser_context);
-    StoragePartition* storage_partition = process_host_->GetStoragePartition();
+        ChromeBlobStorageContext::GetFor(browser_context_);
     std::string uuid = handle.substr(strlen(kBlobPrefix));
     stream =
-        io_context_->OpenBlob(blob_context, storage_partition, handle, uuid);
+        io_context_->OpenBlob(blob_context, storage_partition_, handle, uuid);
   }
 
   if (!stream) {
diff --git a/content/browser/devtools/protocol/io_handler.h b/content/browser/devtools/protocol/io_handler.h
index 3cee43866acee1..7a25cae00ef85d 100644
--- a/content/browser/devtools/protocol/io_handler.h
+++ b/content/browser/devtools/protocol/io_handler.h
@@ -11,7 +11,9 @@
 #include "content/browser/devtools/protocol/io.h"
 
 namespace content {
+class BrowserContext;
 class DevToolsIOContext;
+class StoragePartition;
 
 namespace protocol {
 
@@ -22,7 +24,7 @@ class IOHandler : public DevToolsDomainHandler,
   ~IOHandler() override;
 
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
 
   // Protocol methods.
@@ -41,7 +43,8 @@ class IOHandler : public DevToolsDomainHandler,
 
   std::unique_ptr<IO::Frontend> frontend_;
   DevToolsIOContext* io_context_;
-  RenderProcessHost* process_host_;
+  BrowserContext* browser_context_;
+  StoragePartition* storage_partition_;
   base::WeakPtrFactory<IOHandler> weak_factory_;
 
   DISALLOW_COPY_AND_ASSIGN(IOHandler);
diff --git a/content/browser/devtools/protocol/network_handler.cc b/content/browser/devtools/protocol/network_handler.cc
index a4cf35431c16d6..d1bdbb08b69e37 100644
--- a/content/browser/devtools/protocol/network_handler.cc
+++ b/content/browser/devtools/protocol/network_handler.cc
@@ -694,7 +694,8 @@ void ConfigureServiceWorkerContextOnIO() {
 
 NetworkHandler::NetworkHandler(const std::string& host_id)
     : DevToolsDomainHandler(Network::Metainfo::domainName),
-      process_(nullptr),
+      browser_context_(nullptr),
+      storage_partition_(nullptr),
       host_(nullptr),
       enabled_(false),
       host_id_(host_id),
@@ -724,9 +725,17 @@ void NetworkHandler::Wire(UberDispatcher* dispatcher) {
   Network::Dispatcher::wire(dispatcher, this);
 }
 
-void NetworkHandler::SetRenderer(RenderProcessHost* process_host,
+void NetworkHandler::SetRenderer(int render_process_host_id,
                                  RenderFrameHostImpl* frame_host) {
-  process_ = process_host;
+  RenderProcessHost* process_host =
+      RenderProcessHost::FromID(render_process_host_id);
+  if (process_host) {
+    storage_partition_ = process_host->GetStoragePartition();
+    browser_context_ = process_host->GetBrowserContext();
+  } else {
+    storage_partition_ = nullptr;
+    browser_context_ = nullptr;
+  }
   host_ = frame_host;
 }
 
@@ -775,13 +784,12 @@ class DevtoolsClearCacheObserver
 
 void NetworkHandler::ClearBrowserCache(
     std::unique_ptr<ClearBrowserCacheCallback> callback) {
-  if (!process_) {
+  if (!browser_context_) {
     callback->sendFailure(Response::InternalError());
     return;
   }
   content::BrowsingDataRemover* remover =
-      content::BrowserContext::GetBrowsingDataRemover(
-          process_->GetBrowserContext());
+      content::BrowserContext::GetBrowsingDataRemover(browser_context_);
   remover->RemoveAndReply(
       base::Time(), base::Time::Max(),
       content::BrowsingDataRemover::DATA_TYPE_CACHE,
@@ -791,7 +799,7 @@ void NetworkHandler::ClearBrowserCache(
 
 void NetworkHandler::ClearBrowserCookies(
     std::unique_ptr<ClearBrowserCookiesCallback> callback) {
-  if (!process_) {
+  if (!storage_partition_) {
     callback->sendFailure(Response::InternalError());
     return;
   }
@@ -800,8 +808,7 @@ void NetworkHandler::ClearBrowserCookies(
       BrowserThread::IO, FROM_HERE,
       base::BindOnce(
           &ClearCookiesOnIO,
-          base::Unretained(
-              process_->GetStoragePartition()->GetURLRequestContext()),
+          base::Unretained(storage_partition_->GetURLRequestContext()),
           std::move(callback)));
 }
 
@@ -820,14 +827,12 @@ void NetworkHandler::GetCookies(Maybe<Array<String>> protocol_urls,
       BrowserThread::IO, FROM_HERE,
       base::BindOnce(
           &CookieRetriever::RetrieveCookiesOnIO, retriever,
-          base::Unretained(
-              process_->GetStoragePartition()->GetURLRequestContext()),
-          urls));
+          base::Unretained(storage_partition_->GetURLRequestContext()), urls));
 }
 
 void NetworkHandler::GetAllCookies(
     std::unique_ptr<GetAllCookiesCallback> callback) {
-  if (!process_) {
+  if (!storage_partition_) {
     callback->sendFailure(Response::InternalError());
     return;
   }
@@ -839,8 +844,7 @@ void NetworkHandler::GetAllCookies(
       BrowserThread::IO, FROM_HERE,
       base::BindOnce(
           &CookieRetriever::RetrieveAllCookiesOnIO, retriever,
-          base::Unretained(
-              process_->GetStoragePartition()->GetURLRequestContext())));
+          base::Unretained(storage_partition_->GetURLRequestContext())));
 }
 
 void NetworkHandler::SetCookie(const std::string& name,
@@ -853,7 +857,7 @@ void NetworkHandler::SetCookie(const std::string& name,
                                Maybe<std::string> same_site,
                                Maybe<double> expires,
                                std::unique_ptr<SetCookieCallback> callback) {
-  if (!process_) {
+  if (!storage_partition_) {
     callback->sendFailure(Response::InternalError());
     return;
   }
@@ -867,19 +871,17 @@ void NetworkHandler::SetCookie(const std::string& name,
       BrowserThread::IO, FROM_HERE,
       base::BindOnce(
           &SetCookieOnIO,
-          base::Unretained(
-              process_->GetStoragePartition()->GetURLRequestContext()),
-          name, value, url.fromMaybe(""), domain.fromMaybe(""),
-          path.fromMaybe(""), secure.fromMaybe(false),
-          http_only.fromMaybe(false), same_site.fromMaybe(""),
-          expires.fromMaybe(-1),
+          base::Unretained(storage_partition_->GetURLRequestContext()), name,
+          value, url.fromMaybe(""), domain.fromMaybe(""), path.fromMaybe(""),
+          secure.fromMaybe(false), http_only.fromMaybe(false),
+          same_site.fromMaybe(""), expires.fromMaybe(-1),
           base::BindOnce(&CookieSetOnIO, std::move(callback))));
 }
 
 void NetworkHandler::SetCookies(
     std::unique_ptr<protocol::Array<Network::CookieParam>> cookies,
     std::unique_ptr<SetCookiesCallback> callback) {
-  if (!process_) {
+  if (!storage_partition_) {
     callback->sendFailure(Response::InternalError());
     return;
   }
@@ -888,8 +890,7 @@ void NetworkHandler::SetCookies(
       BrowserThread::IO, FROM_HERE,
       base::BindOnce(
           &SetCookiesOnIO,
-          base::Unretained(
-              process_->GetStoragePartition()->GetURLRequestContext()),
+          base::Unretained(storage_partition_->GetURLRequestContext()),
           std::move(cookies),
           base::BindOnce(&CookiesSetOnIO, std::move(callback))));
 }
@@ -900,7 +901,7 @@ void NetworkHandler::DeleteCookies(
     Maybe<std::string> domain,
     Maybe<std::string> path,
     std::unique_ptr<DeleteCookiesCallback> callback) {
-  if (!process_) {
+  if (!storage_partition_) {
     callback->sendFailure(Response::InternalError());
     return;
   }
@@ -913,9 +914,8 @@ void NetworkHandler::DeleteCookies(
       BrowserThread::IO, FROM_HERE,
       base::BindOnce(
           &DeleteCookiesOnIO,
-          base::Unretained(
-              process_->GetStoragePartition()->GetURLRequestContext()),
-          name, url.fromMaybe(""), domain.fromMaybe(""), path.fromMaybe(""),
+          base::Unretained(storage_partition_->GetURLRequestContext()), name,
+          url.fromMaybe(""), domain.fromMaybe(""), path.fromMaybe(""),
           base::BindOnce(&DeleteCookiesCallback::sendSuccess,
                          std::move(callback))));
 }
@@ -1217,8 +1217,7 @@ void NetworkHandler::ContinueInterceptedRequest(
     Maybe<protocol::Network::AuthChallengeResponse> auth_challenge_response,
     std::unique_ptr<ContinueInterceptedRequestCallback> callback) {
   DevToolsInterceptorController* interceptor =
-      DevToolsInterceptorController::FromBrowserContext(
-          process_->GetBrowserContext());
+      DevToolsInterceptorController::FromBrowserContext(browser_context_);
   if (!interceptor) {
     callback->sendFailure(Response::InternalError());
     return;
@@ -1260,9 +1259,7 @@ void NetworkHandler::GetResponseBodyForInterception(
     const String& interception_id,
     std::unique_ptr<GetResponseBodyForInterceptionCallback> callback) {
   DevToolsInterceptorController* interceptor =
-      DevToolsInterceptorController::FromBrowserContext(
-          process_->GetBrowserContext());
-
+      DevToolsInterceptorController::FromBrowserContext(browser_context_);
   if (!interceptor) {
     callback->sendFailure(Response::InternalError());
     return;
@@ -1400,10 +1397,10 @@ void NetworkHandler::RequestIntercepted(
 
 void NetworkHandler::SetNetworkConditions(
     network::mojom::NetworkConditionsPtr conditions) {
-  if (!process_)
+  if (!storage_partition_)
     return;
-  StoragePartition* partition = process_->GetStoragePartition();
-  network::mojom::NetworkContext* context = partition->GetNetworkContext();
+  network::mojom::NetworkContext* context =
+      storage_partition_->GetNetworkContext();
   context->SetNetworkConditions(host_id_, std::move(conditions));
 }
 
diff --git a/content/browser/devtools/protocol/network_handler.h b/content/browser/devtools/protocol/network_handler.h
index 4a92de40c1875a..e261fb8d083e64 100644
--- a/content/browser/devtools/protocol/network_handler.h
+++ b/content/browser/devtools/protocol/network_handler.h
@@ -28,13 +28,14 @@ struct URLLoaderCompletionStatus;
 }  // namespace network
 
 namespace content {
+class BrowserContext;
 class DevToolsAgentHostImpl;
 class RenderFrameHostImpl;
-struct GlobalRequestID;
 class InterceptionHandle;
 class NavigationHandle;
 class NavigationRequest;
 class NavigationThrottle;
+class StoragePartition;
 struct GlobalRequestID;
 struct InterceptedRequestInfo;
 struct ResourceRequest;
@@ -50,7 +51,7 @@ class NetworkHandler : public DevToolsDomainHandler,
   static std::vector<NetworkHandler*> ForAgentHost(DevToolsAgentHostImpl* host);
 
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int render_process_id,
                    RenderFrameHostImpl* frame_host) override;
 
   Response Enable(Maybe<int> max_total_size,
@@ -150,7 +151,8 @@ class NetworkHandler : public DevToolsDomainHandler,
   void SetNetworkConditions(network::mojom::NetworkConditionsPtr conditions);
 
   std::unique_ptr<Network::Frontend> frontend_;
-  RenderProcessHost* process_;
+  BrowserContext* browser_context_;
+  StoragePartition* storage_partition_;
   RenderFrameHostImpl* host_;
   bool enabled_;
   std::string user_agent_;
diff --git a/content/browser/devtools/protocol/page_handler.cc b/content/browser/devtools/protocol/page_handler.cc
index 3e22c674bacb96..c02c2f4b9f4eae 100644
--- a/content/browser/devtools/protocol/page_handler.cc
+++ b/content/browser/devtools/protocol/page_handler.cc
@@ -148,7 +148,7 @@ std::vector<PageHandler*> PageHandler::ForAgentHost(
       host, Page::Metainfo::domainName);
 }
 
-void PageHandler::SetRenderer(RenderProcessHost* process_host,
+void PageHandler::SetRenderer(int process_host_id,
                               RenderFrameHostImpl* frame_host) {
   if (host_ == frame_host)
     return;
diff --git a/content/browser/devtools/protocol/page_handler.h b/content/browser/devtools/protocol/page_handler.h
index 3ffdea9f5862dd..9c02cad8a812a6 100644
--- a/content/browser/devtools/protocol/page_handler.h
+++ b/content/browser/devtools/protocol/page_handler.h
@@ -61,7 +61,7 @@ class PageHandler : public DevToolsDomainHandler,
   static std::vector<PageHandler*> ForAgentHost(DevToolsAgentHostImpl* host);
 
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
   void OnSwapCompositorFrame(viz::CompositorFrameMetadata frame_metadata);
   void OnSynchronousSwapCompositorFrame(
diff --git a/content/browser/devtools/protocol/security_handler.cc b/content/browser/devtools/protocol/security_handler.cc
index 7494e25fefd7f5..bd4a401fdca89c 100644
--- a/content/browser/devtools/protocol/security_handler.cc
+++ b/content/browser/devtools/protocol/security_handler.cc
@@ -132,7 +132,7 @@ void SecurityHandler::AttachToRenderFrameHost() {
   DidChangeVisibleSecurityState();
 }
 
-void SecurityHandler::SetRenderer(RenderProcessHost* process_host,
+void SecurityHandler::SetRenderer(int process_host_id,
                                   RenderFrameHostImpl* frame_host) {
   host_ = frame_host;
   if (enabled_ && host_)
diff --git a/content/browser/devtools/protocol/security_handler.h b/content/browser/devtools/protocol/security_handler.h
index 30ff780ebe4c79..2e90a92107c3da 100644
--- a/content/browser/devtools/protocol/security_handler.h
+++ b/content/browser/devtools/protocol/security_handler.h
@@ -35,7 +35,7 @@ class SecurityHandler : public DevToolsDomainHandler,
 
   // DevToolsDomainHandler overrides
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
 
   // Security::Backend overrides.
diff --git a/content/browser/devtools/protocol/service_worker_handler.cc b/content/browser/devtools/protocol/service_worker_handler.cc
index 64dff1f030bb3a..7962787ecd5717 100644
--- a/content/browser/devtools/protocol/service_worker_handler.cc
+++ b/content/browser/devtools/protocol/service_worker_handler.cc
@@ -49,8 +49,6 @@ void StatusNoOpKeepingRegistration(
     ServiceWorkerStatusCode status) {
 }
 
-void PushDeliveryNoOp(mojom::PushDeliveryStatus status) {}
-
 const std::string GetVersionRunningStatusString(
     EmbeddedWorkerStatus running_status) {
   switch (running_status) {
@@ -158,7 +156,8 @@ void DispatchSyncEventOnIO(scoped_refptr<ServiceWorkerContextWrapper> context,
 ServiceWorkerHandler::ServiceWorkerHandler()
     : DevToolsDomainHandler(ServiceWorker::Metainfo::domainName),
       enabled_(false),
-      process_(nullptr),
+      browser_context_(nullptr),
+      storage_partition_(nullptr),
       weak_factory_(this) {}
 
 ServiceWorkerHandler::~ServiceWorkerHandler() {
@@ -169,19 +168,21 @@ void ServiceWorkerHandler::Wire(UberDispatcher* dispatcher) {
   ServiceWorker::Dispatcher::wire(dispatcher, this);
 }
 
-void ServiceWorkerHandler::SetRenderer(RenderProcessHost* process_host,
+void ServiceWorkerHandler::SetRenderer(int process_host_id,
                                        RenderFrameHostImpl* frame_host) {
-  process_ = process_host;
+  RenderProcessHost* process_host = RenderProcessHost::FromID(process_host_id);
   // Do not call UpdateHosts yet, wait for load to commit.
   if (!process_host) {
     ClearForceUpdate();
     context_ = nullptr;
     return;
   }
-  StoragePartition* partition = process_host->GetStoragePartition();
-  DCHECK(partition);
+
+  storage_partition_ =
+      static_cast<StoragePartitionImpl*>(process_host->GetStoragePartition());
+  DCHECK(storage_partition_);
   context_ = static_cast<ServiceWorkerContextWrapper*>(
-      partition->GetServiceWorkerContext());
+      storage_partition_->GetServiceWorkerContext());
 }
 
 Response ServiceWorkerHandler::Enable() {
@@ -310,7 +311,7 @@ Response ServiceWorkerHandler::DeliverPushMessage(
     const std::string& data) {
   if (!enabled_)
     return CreateDomainNotEnabledErrorResponse();
-  if (!process_)
+  if (!browser_context_)
     return CreateContextErrorResponse();
   int64_t id = 0;
   if (!base::StringToInt64(registration_id, &id))
@@ -318,9 +319,10 @@ Response ServiceWorkerHandler::DeliverPushMessage(
   PushEventPayload payload;
   if (data.size() > 0)
     payload.setData(data);
-  BrowserContext::DeliverPushMessage(process_->GetBrowserContext(),
-                                     GURL(origin), id, payload,
-                                     base::Bind(&PushDeliveryNoOp));
+  BrowserContext::DeliverPushMessage(
+      browser_context_, GURL(origin), id, payload,
+      base::BindRepeating([](mojom::PushDeliveryStatus status) {}));
+
   return Response::OK();
 }
 
@@ -331,15 +333,14 @@ Response ServiceWorkerHandler::DispatchSyncEvent(
     bool last_chance) {
   if (!enabled_)
     return CreateDomainNotEnabledErrorResponse();
-  if (!process_)
+  if (!storage_partition_)
     return CreateContextErrorResponse();
   int64_t id = 0;
   if (!base::StringToInt64(registration_id, &id))
     return CreateInvalidVersionIdErrorResponse();
 
-  StoragePartitionImpl* partition =
-      static_cast<StoragePartitionImpl*>(process_->GetStoragePartition());
-  BackgroundSyncContext* sync_context = partition->GetBackgroundSyncContext();
+  BackgroundSyncContext* sync_context =
+      storage_partition_->GetBackgroundSyncContext();
 
   BrowserThread::PostTask(BrowserThread::IO, FROM_HERE,
                           base::BindOnce(&DispatchSyncEventOnIO, context_,
diff --git a/content/browser/devtools/protocol/service_worker_handler.h b/content/browser/devtools/protocol/service_worker_handler.h
index 3de2f93cafd2b1..ec9ab86d0ca6d1 100644
--- a/content/browser/devtools/protocol/service_worker_handler.h
+++ b/content/browser/devtools/protocol/service_worker_handler.h
@@ -20,9 +20,11 @@
 
 namespace content {
 
+class BrowserContext;
 class RenderFrameHostImpl;
 class ServiceWorkerContextWatcher;
 class ServiceWorkerContextWrapper;
+class StoragePartitionImpl;
 
 namespace protocol {
 
@@ -33,7 +35,7 @@ class ServiceWorkerHandler : public DevToolsDomainHandler,
   ~ServiceWorkerHandler() override;
 
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
 
   Response Enable() override;
@@ -71,7 +73,8 @@ class ServiceWorkerHandler : public DevToolsDomainHandler,
   std::unique_ptr<ServiceWorker::Frontend> frontend_;
   bool enabled_;
   scoped_refptr<ServiceWorkerContextWatcher> context_watcher_;
-  RenderProcessHost* process_;
+  BrowserContext* browser_context_;
+  StoragePartitionImpl* storage_partition_;
 
   base::WeakPtrFactory<ServiceWorkerHandler> weak_factory_;
 
diff --git a/content/browser/devtools/protocol/storage_handler.cc b/content/browser/devtools/protocol/storage_handler.cc
index 9e8d8d20edc28f..29eb1b7e8470dd 100644
--- a/content/browser/devtools/protocol/storage_handler.cc
+++ b/content/browser/devtools/protocol/storage_handler.cc
@@ -239,7 +239,7 @@ class StorageHandler::IndexedDBObserver : IndexedDBContextImpl::Observer {
 
 StorageHandler::StorageHandler()
     : DevToolsDomainHandler(Storage::Metainfo::domainName),
-      process_(nullptr),
+      storage_partition_(nullptr),
       weak_ptr_factory_(this) {}
 
 StorageHandler::~StorageHandler() {
@@ -252,9 +252,10 @@ void StorageHandler::Wire(UberDispatcher* dispatcher) {
   Storage::Dispatcher::wire(dispatcher, this);
 }
 
-void StorageHandler::SetRenderer(RenderProcessHost* process_host,
+void StorageHandler::SetRenderer(int process_host_id,
                                  RenderFrameHostImpl* frame_host) {
-  process_ = process_host;
+  RenderProcessHost* process = RenderProcessHost::FromID(process_host_id);
+  storage_partition_ = process ? process->GetStoragePartition() : nullptr;
 }
 
 Response StorageHandler::Disable() {
@@ -276,10 +277,9 @@ void StorageHandler::ClearDataForOrigin(
     const std::string& origin,
     const std::string& storage_types,
     std::unique_ptr<ClearDataForOriginCallback> callback) {
-  if (!process_)
+  if (!storage_partition_)
     return callback->sendFailure(Response::InternalError());
 
-  StoragePartition* partition = process_->GetStoragePartition();
   std::vector<std::string> types = base::SplitString(
       storage_types, ",", base::TRIM_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
   std::unordered_set<std::string> set(types.begin(), types.end());
@@ -310,18 +310,18 @@ void StorageHandler::ClearDataForOrigin(
         Response::InvalidParams("No valid storage type specified"));
   }
 
-  partition->ClearData(remove_mask,
-                       StoragePartition::QUOTA_MANAGED_STORAGE_MASK_ALL,
-                       GURL(origin), StoragePartition::OriginMatcherFunction(),
-                       base::Time(), base::Time::Max(),
-                       base::BindOnce(&ClearDataForOriginCallback::sendSuccess,
-                                      std::move(callback)));
+  storage_partition_->ClearData(
+      remove_mask, StoragePartition::QUOTA_MANAGED_STORAGE_MASK_ALL,
+      GURL(origin), StoragePartition::OriginMatcherFunction(), base::Time(),
+      base::Time::Max(),
+      base::BindOnce(&ClearDataForOriginCallback::sendSuccess,
+                     std::move(callback)));
 }
 
 void StorageHandler::GetUsageAndQuota(
     const String& origin,
     std::unique_ptr<GetUsageAndQuotaCallback> callback) {
-  if (!process_)
+  if (!storage_partition_)
     return callback->sendFailure(Response::InternalError());
 
   GURL origin_url(origin);
@@ -330,8 +330,7 @@ void StorageHandler::GetUsageAndQuota(
         Response::Error(origin + " is not a valid URL"));
   }
 
-  storage::QuotaManager* manager =
-      process_->GetStoragePartition()->GetQuotaManager();
+  storage::QuotaManager* manager = storage_partition_->GetQuotaManager();
   BrowserThread::PostTask(
       BrowserThread::IO, FROM_HERE,
       base::BindOnce(&GetUsageAndQuotaOnIOThread, base::RetainedRef(manager),
@@ -339,7 +338,7 @@ void StorageHandler::GetUsageAndQuota(
 }
 
 Response StorageHandler::TrackCacheStorageForOrigin(const std::string& origin) {
-  if (!process_)
+  if (!storage_partition_)
     return Response::InternalError();
 
   GURL origin_url(origin);
@@ -356,7 +355,7 @@ Response StorageHandler::TrackCacheStorageForOrigin(const std::string& origin) {
 
 Response StorageHandler::UntrackCacheStorageForOrigin(
     const std::string& origin) {
-  if (!process_)
+  if (!storage_partition_)
     return Response::InternalError();
 
   GURL origin_url(origin);
@@ -372,7 +371,7 @@ Response StorageHandler::UntrackCacheStorageForOrigin(
 }
 
 Response StorageHandler::TrackIndexedDBForOrigin(const std::string& origin) {
-  if (!process_)
+  if (!storage_partition_)
     return Response::InternalError();
 
   GURL origin_url(origin);
@@ -387,7 +386,7 @@ Response StorageHandler::TrackIndexedDBForOrigin(const std::string& origin) {
 }
 
 Response StorageHandler::UntrackIndexedDBForOrigin(const std::string& origin) {
-  if (!process_)
+  if (!storage_partition_)
     return Response::InternalError();
 
   GURL origin_url(origin);
@@ -408,7 +407,7 @@ StorageHandler::GetCacheStorageObserver() {
     cache_storage_observer_ = std::make_unique<CacheStorageObserver>(
         weak_ptr_factory_.GetWeakPtr(),
         static_cast<CacheStorageContextImpl*>(
-            process_->GetStoragePartition()->GetCacheStorageContext()));
+            storage_partition_->GetCacheStorageContext()));
   }
   return cache_storage_observer_.get();
 }
@@ -419,7 +418,7 @@ StorageHandler::IndexedDBObserver* StorageHandler::GetIndexedDBObserver() {
     indexed_db_observer_ = std::make_unique<IndexedDBObserver>(
         weak_ptr_factory_.GetWeakPtr(),
         static_cast<IndexedDBContextImpl*>(
-            process_->GetStoragePartition()->GetIndexedDBContext()));
+            storage_partition_->GetIndexedDBContext()));
   }
   return indexed_db_observer_.get();
 }
diff --git a/content/browser/devtools/protocol/storage_handler.h b/content/browser/devtools/protocol/storage_handler.h
index dc33e453480e79..d9df1114abd5cc 100644
--- a/content/browser/devtools/protocol/storage_handler.h
+++ b/content/browser/devtools/protocol/storage_handler.h
@@ -17,6 +17,8 @@
 #include "content/browser/indexed_db/indexed_db_context_impl.h"
 
 namespace content {
+class StoragePartition;
+
 namespace protocol {
 
 class StorageHandler : public DevToolsDomainHandler,
@@ -27,7 +29,7 @@ class StorageHandler : public DevToolsDomainHandler,
 
   // content::protocol::DevToolsDomainHandler
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
   Response Disable() override;
 
@@ -64,7 +66,7 @@ class StorageHandler : public DevToolsDomainHandler,
                                      const base::string16& object_store_name);
 
   std::unique_ptr<Storage::Frontend> frontend_;
-  RenderProcessHost* process_;
+  StoragePartition* storage_partition_;
   std::unique_ptr<CacheStorageObserver> cache_storage_observer_;
   std::unique_ptr<IndexedDBObserver> indexed_db_observer_;
 
diff --git a/content/browser/devtools/protocol/target_handler.cc b/content/browser/devtools/protocol/target_handler.cc
index a4d405f837bd2a..bdcbf7f1b5e1ed 100644
--- a/content/browser/devtools/protocol/target_handler.cc
+++ b/content/browser/devtools/protocol/target_handler.cc
@@ -208,7 +208,7 @@ void TargetHandler::Wire(UberDispatcher* dispatcher) {
   Target::Dispatcher::wire(dispatcher, this);
 }
 
-void TargetHandler::SetRenderer(RenderProcessHost* process_host,
+void TargetHandler::SetRenderer(int process_host_id,
                                 RenderFrameHostImpl* frame_host) {
   auto_attacher_.SetRenderFrameHost(frame_host);
 }
diff --git a/content/browser/devtools/protocol/target_handler.h b/content/browser/devtools/protocol/target_handler.h
index d9d4a8955aac34..21adcd6a529a1a 100644
--- a/content/browser/devtools/protocol/target_handler.h
+++ b/content/browser/devtools/protocol/target_handler.h
@@ -34,7 +34,7 @@ class TargetHandler : public DevToolsDomainHandler,
   static std::vector<TargetHandler*> ForAgentHost(DevToolsAgentHostImpl* host);
 
   void Wire(UberDispatcher* dispatcher) override;
-  void SetRenderer(RenderProcessHost* process_host,
+  void SetRenderer(int process_host_id,
                    RenderFrameHostImpl* frame_host) override;
   Response Disable() override;
 
diff --git a/content/browser/devtools/render_frame_devtools_agent_host.cc b/content/browser/devtools/render_frame_devtools_agent_host.cc
index 4cbceb60293ca9..f8ac27bd19d025 100644
--- a/content/browser/devtools/render_frame_devtools_agent_host.cc
+++ b/content/browser/devtools/render_frame_devtools_agent_host.cc
@@ -292,7 +292,8 @@ WebContents* RenderFrameDevToolsAgentHost::GetWebContents() {
 }
 
 void RenderFrameDevToolsAgentHost::AttachSession(DevToolsSession* session) {
-  session->SetRenderer(frame_host_ ? frame_host_->GetProcess() : nullptr,
+  session->SetRenderer(frame_host_ ? frame_host_->GetProcess()->GetID()
+                                   : ChildProcessHost::kInvalidUniqueID,
                        frame_host_);
 
   protocol::EmulationHandler* emulation_handler =
@@ -441,7 +442,7 @@ void RenderFrameDevToolsAgentHost::UpdateFrameHost(
   if (IsAttached()) {
     GrantPolicy();
     for (DevToolsSession* session : sessions()) {
-      session->SetRenderer(frame_host ? frame_host->GetProcess() : nullptr,
+      session->SetRenderer(frame_host ? frame_host->GetProcess()->GetID() : -1,
                            frame_host);
     }
     MaybeReattachToRenderFrame();
diff --git a/content/browser/devtools/service_worker_devtools_agent_host.cc b/content/browser/devtools/service_worker_devtools_agent_host.cc
index cc0517ade4a317..3cea99c268ce18 100644
--- a/content/browser/devtools/service_worker_devtools_agent_host.cc
+++ b/content/browser/devtools/service_worker_devtools_agent_host.cc
@@ -122,10 +122,7 @@ void ServiceWorkerDevToolsAgentHost::AttachSession(DevToolsSession* session) {
                               base::BindOnce(&SetDevToolsAttachedOnIO,
                                              context_weak_, version_id_, true));
     }
-    // RenderProcessHost should not be null here, but even if it _is_ null,
-    // session does not depend on the process to do messaging.
-    session->SetRenderer(RenderProcessHost::FromID(worker_process_id_),
-                         nullptr);
+    session->SetRenderer(worker_process_id_, nullptr);
     session->AttachToAgent(agent_ptr_);
   }
   session->AddHandler(base::WrapUnique(new protocol::InspectorHandler()));
@@ -159,11 +156,8 @@ void ServiceWorkerDevToolsAgentHost::WorkerReadyForInspection(
                                            context_weak_, version_id_, true));
   }
 
-  // RenderProcessHost should not be null here, but even if it _is_ null,
-  // session does not depend on the process to do messaging.
-  RenderProcessHost* host = RenderProcessHost::FromID(worker_process_id_);
   for (DevToolsSession* session : sessions()) {
-    session->SetRenderer(host, nullptr);
+    session->SetRenderer(worker_process_id_, nullptr);
     session->AttachToAgent(agent_ptr_);
   }
 }
@@ -174,9 +168,8 @@ void ServiceWorkerDevToolsAgentHost::WorkerRestarted(int worker_process_id,
   state_ = WORKER_NOT_READY;
   worker_process_id_ = worker_process_id;
   worker_route_id_ = worker_route_id;
-  RenderProcessHost* host = RenderProcessHost::FromID(worker_process_id_);
   for (DevToolsSession* session : sessions())
-    session->SetRenderer(host, nullptr);
+    session->SetRenderer(worker_process_id_, nullptr);
 }
 
 void ServiceWorkerDevToolsAgentHost::WorkerDestroyed() {
@@ -186,7 +179,7 @@ void ServiceWorkerDevToolsAgentHost::WorkerDestroyed() {
   for (auto* inspector : protocol::InspectorHandler::ForAgentHost(this))
     inspector->TargetCrashed();
   for (DevToolsSession* session : sessions())
-    session->SetRenderer(nullptr, nullptr);
+    session->SetRenderer(-1, nullptr);
 }
 
 }  // namespace content
diff --git a/content/browser/devtools/shared_worker_devtools_agent_host.cc b/content/browser/devtools/shared_worker_devtools_agent_host.cc
index a93ff3a7ea75b9..3fb53685c6bee6 100644
--- a/content/browser/devtools/shared_worker_devtools_agent_host.cc
+++ b/content/browser/devtools/shared_worker_devtools_agent_host.cc
@@ -34,7 +34,10 @@ SharedWorkerDevToolsAgentHost::~SharedWorkerDevToolsAgentHost() {
 }
 
 BrowserContext* SharedWorkerDevToolsAgentHost::GetBrowserContext() {
-  RenderProcessHost* rph = GetProcess();
+  if (!worker_host_)
+    return nullptr;
+  RenderProcessHost* rph =
+      RenderProcessHost::FromID(worker_host_->process_id());
   return rph ? rph->GetBrowserContext() : nullptr;
 }
 
@@ -67,7 +70,7 @@ void SharedWorkerDevToolsAgentHost::AttachSession(DevToolsSession* session) {
   session->AddHandler(std::make_unique<protocol::InspectorHandler>());
   session->AddHandler(std::make_unique<protocol::NetworkHandler>(GetId()));
   session->AddHandler(std::make_unique<protocol::SchemaHandler>());
-  session->SetRenderer(GetProcess(), nullptr);
+  session->SetRenderer(worker_host_ ? worker_host_->process_id() : -1, nullptr);
   if (state_ == WORKER_READY)
     session->AttachToAgent(EnsureAgent());
 }
@@ -101,7 +104,7 @@ void SharedWorkerDevToolsAgentHost::WorkerRestarted(
   state_ = WORKER_NOT_READY;
   worker_host_ = worker_host;
   for (DevToolsSession* session : sessions())
-    session->SetRenderer(GetProcess(), nullptr);
+    session->SetRenderer(worker_host_->process_id(), nullptr);
 }
 
 void SharedWorkerDevToolsAgentHost::WorkerDestroyed() {
@@ -111,16 +114,11 @@ void SharedWorkerDevToolsAgentHost::WorkerDestroyed() {
   for (auto* inspector : protocol::InspectorHandler::ForAgentHost(this))
     inspector->TargetCrashed();
   for (DevToolsSession* session : sessions())
-    session->SetRenderer(nullptr, nullptr);
+    session->SetRenderer(-1, nullptr);
   worker_host_ = nullptr;
   agent_ptr_.reset();
 }
 
-RenderProcessHost* SharedWorkerDevToolsAgentHost::GetProcess() {
-  return worker_host_ ? RenderProcessHost::FromID(worker_host_->process_id())
-                      : nullptr;
-}
-
 const blink::mojom::DevToolsAgentAssociatedPtr&
 SharedWorkerDevToolsAgentHost::EnsureAgent() {
   DCHECK_EQ(WORKER_READY, state_);
diff --git a/content/browser/devtools/shared_worker_devtools_agent_host.h b/content/browser/devtools/shared_worker_devtools_agent_host.h
index 7f4f175d722b3f..15c5de57b5975d 100644
--- a/content/browser/devtools/shared_worker_devtools_agent_host.h
+++ b/content/browser/devtools/shared_worker_devtools_agent_host.h
@@ -14,7 +14,6 @@ namespace content {
 
 class SharedWorkerInstance;
 class SharedWorkerHost;
-class RenderProcessHost;
 
 class SharedWorkerDevToolsAgentHost : public DevToolsAgentHostImpl {
  public:
@@ -50,7 +49,6 @@ class SharedWorkerDevToolsAgentHost : public DevToolsAgentHostImpl {
 
  private:
   ~SharedWorkerDevToolsAgentHost() override;
-  RenderProcessHost* GetProcess();
   const blink::mojom::DevToolsAgentAssociatedPtr& EnsureAgent();
 
   enum WorkerState {
